File: dsh-code-runtime/src/main/resources/run_code_shim.py
#!/usr/bin/env python3
# run_code 运行时 shim (Python) — 加载模型写的 Python 程序，提供 tools 绑定（行分隔 JSON-RPC over stdio）
import asyncio
import io
import json
import os
import sys
import threading

CODE_FILE = sys.argv[1]
with open(CODE_FILE, encoding="utf-8") as f:
    CODE = f.read()

# 协议写入走原始 fd，避免与程序 stdout 捕获冲突
REAL_STDOUT = os.fdopen(os.dup(1), "w", encoding="utf-8")
PROTOCOL_FD = os.dup(1)


def proto(obj):
    os.write(PROTOCOL_FD, (json.dumps(obj) + "\n").encode("utf-8"))


loop = asyncio.new_event_loop()
pending = {}
next_id = [1]


def reader_thread():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
        except Exception:
            continue
        rid = msg.get("id")
        if rid is not None and rid in pending:
            fut = pending.pop(rid)
            if msg.get("error"):
                loop.call_soon_threadsafe(fut.set_exception, RuntimeError(msg["error"]))
            else:
                loop.call_soon_threadsafe(fut.set_result, msg.get("result"))


threading.Thread(target=reader_thread, daemon=True).start()


async def request(name, args):
    rid = next_id[0]
    next_id[0] += 1
    fut = loop.create_future()
    pending[rid] = fut
    proto({"id": rid, "name": name, "args": args or {}})
    return await fut


class Tools:
    def __getattr__(self, name):
        def call(args=None):
            return request(name, args or {})
        return call


class LogCapture:
    def __init__(self):
        self.buffer = io.StringIO()

    def write(self, s):
        self.buffer.write(s)
        REAL_STDOUT.write(s)
        return len(s)

    def flush(self):
        REAL_STDOUT.flush()


logs_capture = LogCapture()
orig_stdout = sys.stdout
sys.stdout = logs_capture


def finish(payload, exit_code):
    logs_capture.flush()
    os.write(PROTOCOL_FD, ("__DSH_RESULT__" + json.dumps(payload) + "\n").encode("utf-8"))
    try:
        os.close(PROTOCOL_FD)
    except Exception:
        pass
    # os._exit 跳过一切异常处理/finally，避免二次 finish 崩溃
    os._exit(exit_code)


async def main_wrapper():
    ns = {"tools": Tools()}
    exec("async def __run():\n" + "\n".join("    " + ln for ln in CODE.splitlines()), ns)
    return await ns["__run"]()


try:
    result = loop.run_until_complete(main_wrapper())
    finish({"logs": logs_capture.buffer.getvalue().splitlines(), "result": result}, 0)
except Exception as e:  # 程序异常作为错误返回；SystemExit/KeyboardInterrupt 不拦截
    finish({"logs": logs_capture.buffer.getvalue().splitlines(), "error": str(e)}, 1)

File: dsh-code-runtime/src/main/resources/run_code_shim.js
// run_code 运行时 shim — 加载模型写的 JS 程序，提供 tools 绑定（行分隔 JSON-RPC over stdio）
const fs = require('fs');
const readline = require('readline');

const codeFile = process.argv[2];
const code = fs.readFileSync(codeFile, 'utf8');

const rl = readline.createInterface({ input: process.stdin });
const pending = new Map();
let nextId = 1;

rl.on('line', (line) => {
  if (!line) return;
  let msg;
  try {
    msg = JSON.parse(line);
  } catch (e) {
    return;
  }
  if (msg.id !== undefined && pending.has(msg.id)) {
    const { resolve, reject } = pending.get(msg.id);
    pending.delete(msg.id);
    if (msg.error) reject(new Error(msg.error));
    else resolve(msg.result);
  }
});

function callTool(name, args) {
  return new Promise((resolve, reject) => {
    const id = nextId++;
    pending.set(id, { resolve, reject });
    process.stdout.write(JSON.stringify({ id, name, args: args || {} }) + '\n');
  });
}

const tools = new Proxy({}, {
  get: (_, name) => (args) => callTool(name, args || {})
});

const logs = [];
const origLog = console.log;
console.log = (...a) => { logs.push(a.map(String).join(' ')); };

function finish(payload, exitCode) {
  process.stdout.write('__DSH_RESULT__' + JSON.stringify(payload) + '\n');
  process.exit(exitCode);
}

(async () => {
  const fn = new Function('tools', 'console', 'return (async () => { ' + code + '\n })()');
  const result = await fn(tools, console);
  finish({ logs, result: result === undefined ? null : result }, 0);
})().catch((e) => {
  finish({ logs, error: String((e && e.message) || e) }, 1);
});
